File: packages/py/src/airlock/application/agent.py
"""The agent loop with the approval gate built in."""

import json
from collections.abc import Mapping, Sequence
from dataclasses import dataclass

from airlock.application.ports.audit_sink import AuditSink
from airlock.application.ports.clock import Clock
from airlock.application.ports.event_bus import EventPublisher
from airlock.application.ports.gate_policy import GateDecisionInput, GatePolicy
from airlock.application.ports.id_generator import IdGenerator
from airlock.application.ports.llm_provider import CompletionRequest, CompletionResult, LlmProvider
from airlock.application.ports.run_store import RunStore
from airlock.domain.approval import (
    ApprovalDecision,
    ApprovalRequest,
    ApproveDecision,
    EditDecision,
    RejectDecision,
)
from airlock.domain.audit import AuditEvent, AuditEventType
from airlock.domain.conversation import AssistantMessage, ToolMessage, UserMessage
from airlock.domain.errors import RunNotFoundError, UnknownToolError
from airlock.domain.events import EventTopic, RunCompletedEvent
from airlock.domain.identifiers import RequestId, RunId, ToolCallId
from airlock.domain.run import RunState, RunStatus
from airlock.domain.tool import RiskTier, Tool, ToolCall, ToolDefinition

_REJECTION_NOTICE_PREFIX = "Rejected by a human:"


@dataclass(frozen=True)
class AgentDependencies:
    """Everything the Agent needs, injected as one object."""

    provider: LlmProvider
    tools: Sequence[Tool]
    events: EventPublisher
    store: RunStore
    audit: AuditSink
    clock: Clock
    ids: IdGenerator
    gate_policy: GatePolicy
    system_prompt: str


class Agent:
    """Runs a tool-use loop with the gate. Safe tools run automatically; a
    sensitive tool call suspends for a human decision and resumes when one
    arrives. See docs/design/contracts.md for the invariants."""

    def __init__(self, deps: AgentDependencies) -> None:
        self._provider = deps.provider
        self._events = deps.events
        self._store = deps.store
        self._audit = deps.audit
        self._clock = deps.clock
        self._ids = deps.ids
        self._gate_policy = deps.gate_policy
        self._system_prompt = deps.system_prompt
        self._tools: dict[str, Tool] = {tool.name: tool for tool in deps.tools}
        self._tool_definitions: tuple[ToolDefinition, ...] = tuple(
            tool.definition() for tool in deps.tools
        )

    async def run(self, user_input: str) -> RunState:
        """Start a run and advance until it completes or suspends for approval."""
        return await self._advance(self._create_run(user_input))

    async def resume(
        self, run_id: RunId, request_id: RequestId, decision: ApprovalDecision
    ) -> RunState:
        """Apply a human decision and continue. Idempotent, keyed by request_id."""
        state = await self._store.load(run_id)
        if state is None:
            raise RunNotFoundError(run_id)
        request = state.approval
        if not self._is_awaiting(state, request_id) or request is None:
            return state
        await self._apply_decision(state, request, decision)
        state.approval = None
        state.status = RunStatus.RUNNING
        state.cursor += 1
        return await self._advance(state)

    def _create_run(self, user_input: str) -> RunState:
        return RunState(
            run_id=self._ids.run_id(),
            status=RunStatus.RUNNING,
            messages=[UserMessage(content=user_input)],
            pending_tool_calls=(),
            cursor=0,
            approval=None,
            metadata={},
        )

    async def _advance(self, state: RunState) -> RunState:
        while True:
            if await self._process_pending(state):
                return state
            completion = await self._call_model(state)
            if not completion.tool_calls:
                return await self._complete_run(state, completion.text)
            self._start_turn(state, completion)

    async def _process_pending(self, state: RunState) -> bool:
        while state.cursor < len(state.pending_tool_calls):
            call = state.pending_tool_calls[state.cursor]
            tool = self._require_tool(call.name)
            if self._gate_policy.requires_approval(GateDecisionInput(tool, call, state)):
                await self._suspend_for_approval(state, call, tool.risk)
                return True
            await self._run_tool_call(state, tool, call.args, call.id)
            state.cursor += 1
        return False

    async def _suspend_for_approval(
        self, state: RunState, tool_call: ToolCall, risk: RiskTier
    ) -> None:
        request = ApprovalRequest(
            run_id=state.run_id,
            request_id=self._ids.request_id(),
            tool_call=tool_call,
            risk=risk,
            context=state.metadata,
        )
        state.approval = request
        state.status = RunStatus.AWAITING_APPROVAL
        await self._store.save(state)
        await self._record_audit(
            state.run_id,
            AuditEventType.APPROVAL_REQUESTED,
            {"request_id": request.request_id, "tool": tool_call.name},
        )
        await self._events.publish(EventTopic.APPROVAL_REQUESTED, request)

    async def _apply_decision(
        self, state: RunState, request: ApprovalRequest, decision: ApprovalDecision
    ) -> None:
        await self._record_audit(
            state.run_id,
            AuditEventType.APPROVAL_DECIDED,
            {
                "request_id": request.request_id,
                "decision": decision.type,
                "approver": decision.approver,
            },
        )
        match decision:
            case RejectDecision():
                self._append_rejection(state, request.tool_call.id, decision.reason)
                return
            case EditDecision():
                args: Mapping[str, object] = decision.edited_args
            case ApproveDecision():
                args = request.tool_call.args
        tool = self._require_tool(request.tool_call.name)
        await self._run_tool_call(state, tool, args, request.tool_call.id)

    async def _run_tool_call(
        self, state: RunState, tool: Tool, args: Mapping[str, object], tool_call_id: ToolCallId
    ) -> None:
        result = await tool.handler(args)
        await self._record_audit(state.run_id, AuditEventType.TOOL_EXECUTED, {"tool": tool.name})
        state.messages.append(ToolMessage(tool_call_id=tool_call_id, content=_serialize(result)))

    async def _call_model(self, state: RunState) -> CompletionResult:
        completion = await self._provider.complete(
            CompletionRequest(
                system=self._system_prompt,
                messages=state.messages,
                tools=self._tool_definitions,
            )
        )
        await self._record_audit(state.run_id, AuditEventType.MODEL_CALLED, {})
        return completion

    def _start_turn(self, state: RunState, completion: CompletionResult) -> None:
        state.messages.append(
            AssistantMessage(content=completion.text or "", tool_calls=completion.tool_calls)
        )
        state.pending_tool_calls = completion.tool_calls
        state.cursor = 0

    async def _complete_run(self, state: RunState, text: str | None) -> RunState:
        state.messages.append(AssistantMessage(content=text or "", tool_calls=()))
        state.status = RunStatus.COMPLETED
        await self._store.save(state)
        await self._record_audit(state.run_id, AuditEventType.RUN_COMPLETED, {})
        await self._events.publish(EventTopic.RUN_COMPLETED, RunCompletedEvent(run_id=state.run_id))
        return state

    def _append_rejection(self, state: RunState, tool_call_id: ToolCallId, reason: str) -> None:
        state.messages.append(
            ToolMessage(tool_call_id=tool_call_id, content=f"{_REJECTION_NOTICE_PREFIX} {reason}")
        )

    def _is_awaiting(self, state: RunState, request_id: RequestId) -> bool:
        return (
            state.status == RunStatus.AWAITING_APPROVAL
            and state.approval is not None
            and state.approval.request_id == request_id
        )

    def _require_tool(self, name: str) -> Tool:
        tool = self._tools.get(name)
        if tool is None:
            raise UnknownToolError(name)
        return tool

    async def _record_audit(
        self, run_id: RunId, event_type: AuditEventType, data: Mapping[str, object]
    ) -> None:
        event = AuditEvent(
            run_id=run_id, type=event_type, at=self._clock.now().isoformat(), data=dict(data)
        )
        await self._audit.record(event)


def _serialize(result: object) -> str:
    if isinstance(result, str):
        return result
    return json.dumps(result)

File: packages/ts/src/application/agent.ts
import { type ApprovalDecision, type ApprovalRequest, DecisionType } from "../domain/approval";
import { AuditEventType } from "../domain/audit";
import { MessageRole } from "../domain/conversation";
import { EventTopic } from "../domain/events";
import { RunNotFoundError, UnknownToolError } from "../domain/errors";
import type { RequestId, RunId, ToolCallId } from "../domain/identifiers";
import { type RunState, RunStatus } from "../domain/run";
import type { RiskTier, Tool, ToolCall, ToolDefinition } from "../domain/tool";
import type { AuditSink } from "./ports/audit-sink";
import type { Clock } from "./ports/clock";
import type { EventPublisher } from "./ports/event-bus";
import type { GatePolicy } from "./ports/gate-policy";
import type { IdGenerator } from "./ports/id-generator";
import type { CompletionResult, LlmProvider } from "./ports/llm-provider";
import type { RunStore } from "./ports/run-store";

/** Prefix of the tool message recorded when a human rejects an action. */
const REJECTION_NOTICE_PREFIX = "Rejected by a human:";

/** Everything the Agent needs, injected as one object (constructor injection). */
export interface AgentDependencies {
  readonly provider: LlmProvider;
  readonly tools: readonly Tool[];
  readonly events: EventPublisher;
  readonly store: RunStore;
  readonly audit: AuditSink;
  readonly clock: Clock;
  readonly ids: IdGenerator;
  readonly gatePolicy: GatePolicy;
  readonly systemPrompt: string;
}

/**
 * The agent loop with the approval gate built in. Safe tools run automatically;
 * a sensitive tool call suspends the run for a human decision and resumes when
 * one arrives. See docs/design/contracts.md for the invariants.
 */
export class Agent {
  private readonly provider: LlmProvider;
  private readonly events: EventPublisher;
  private readonly store: RunStore;
  private readonly audit: AuditSink;
  private readonly clock: Clock;
  private readonly ids: IdGenerator;
  private readonly gatePolicy: GatePolicy;
  private readonly systemPrompt: string;
  private readonly tools: ReadonlyMap<string, Tool>;
  private readonly toolDefinitions: readonly ToolDefinition[];

  constructor(deps: AgentDependencies) {
    this.provider = deps.provider;
    this.events = deps.events;
    this.store = deps.store;
    this.audit = deps.audit;
    this.clock = deps.clock;
    this.ids = deps.ids;
    this.gatePolicy = deps.gatePolicy;
    this.systemPrompt = deps.systemPrompt;
    this.tools = new Map(deps.tools.map((tool) => [tool.name, tool]));
    this.toolDefinitions = deps.tools.map((tool) => toDefinition(tool));
  }

  /** Start a run and advance until it completes or suspends for approval. */
  async run(input: string): Promise<RunState> {
    return this.advance(this.createRun(input));
  }

  /** Apply a human decision and continue. Idempotent, keyed by `requestId`. */
  async resume(runId: RunId, requestId: RequestId, decision: ApprovalDecision): Promise<RunState> {
    const state = await this.store.load(runId);
    if (state === null) {
      throw new RunNotFoundError(runId);
    }
    const request = state.approval;
    if (!this.isAwaiting(state, requestId) || request === null) {
      return state;
    }
    await this.applyDecision(state, request, decision);
    state.approval = null;
    state.status = RunStatus.Running;
    state.cursor += 1;
    return this.advance(state);
  }

  private createRun(input: string): RunState {
    return {
      runId: this.ids.runId(),
      status: RunStatus.Running,
      messages: [{ role: MessageRole.User, content: input }],
      pendingToolCalls: [],
      cursor: 0,
      approval: null,
      metadata: {},
    };
  }

  private async advance(state: RunState): Promise<RunState> {
    for (;;) {
      if (await this.processPending(state)) {
        return state;
      }
      const completion = await this.callModel(state);
      if (completion.toolCalls.length === 0) {
        return this.completeRun(state, completion.text);
      }
      this.startTurn(state, completion);
    }
  }

  private async processPending(state: RunState): Promise<boolean> {
    while (state.cursor < state.pendingToolCalls.length) {
      const call = state.pendingToolCalls[state.cursor];
      if (call === undefined) {
        break;
      }
      const tool = this.requireTool(call.name);
      if (this.gatePolicy.requiresApproval({ tool, toolCall: call, state })) {
        await this.suspendForApproval(state, call, tool.risk);
        return true;
      }
      await this.runToolCall(state, tool, call.args, call.id);
      state.cursor += 1;
    }
    return false;
  }

  private async suspendForApproval(
    state: RunState,
    toolCall: ToolCall,
    risk: RiskTier,
  ): Promise<void> {
    const request: ApprovalRequest = {
      runId: state.runId,
      requestId: this.ids.requestId(),
      toolCall,
      risk,
      context: state.metadata,
    };
    state.approval = request;
    state.status = RunStatus.AwaitingApproval;
    await this.store.save(state);
    await this.recordAudit(state.runId, AuditEventType.ApprovalRequested, {
      requestId: request.requestId,
      tool: toolCall.name,
    });
    await this.events.publish(EventTopic.ApprovalRequested, request);
  }

  private async applyDecision(
    state: RunState,
    request: ApprovalRequest,
    decision: ApprovalDecision,
  ): Promise<void> {
    await this.recordAudit(state.runId, AuditEventType.ApprovalDecided, {
      requestId: request.requestId,
      decision: decision.type,
      approver: decision.approver,
    });
    if (decision.type === DecisionType.Reject) {
      this.appendRejection(state, request.toolCall.id, decision.reason);
      return;
    }
    const tool = this.requireTool(request.toolCall.name);
    const args = decision.type === DecisionType.Edit ? decision.editedArgs : request.toolCall.args;
    await this.runToolCall(state, tool, args, request.toolCall.id);
  }

  private async runToolCall(
    state: RunState,
    tool: Tool,
    args: Readonly<Record<string, unknown>>,
    toolCallId: ToolCallId,
  ): Promise<void> {
    const result = await tool.handler(args);
    await this.recordAudit(state.runId, AuditEventType.ToolExecuted, { tool: tool.name });
    state.messages.push({ role: MessageRole.Tool, toolCallId, content: serialize(result) });
  }

  private async callModel(state: RunState): Promise<CompletionResult> {
    const completion = await this.provider.complete({
      system: this.systemPrompt,
      messages: state.messages,
      tools: this.toolDefinitions,
    });
    await this.recordAudit(state.runId, AuditEventType.ModelCalled, {});
    return completion;
  }

  private startTurn(state: RunState, completion: CompletionResult): void {
    state.messages.push({
      role: MessageRole.Assistant,
      content: completion.text ?? "",
      toolCalls: completion.toolCalls,
    });
    state.pendingToolCalls = completion.toolCalls;
    state.cursor = 0;
  }

  private async completeRun(state: RunState, text: string | null): Promise<RunState> {
    state.messages.push({ role: MessageRole.Assistant, content: text ?? "", toolCalls: [] });
    state.status = RunStatus.Completed;
    await this.store.save(state);
    await this.recordAudit(state.runId, AuditEventType.RunCompleted, {});
    await this.events.publish(EventTopic.RunCompleted, { runId: state.runId });
    return state;
  }

  private appendRejection(state: RunState, toolCallId: ToolCallId, reason: string): void {
    state.messages.push({
      role: MessageRole.Tool,
      toolCallId,
      content: `${REJECTION_NOTICE_PREFIX} ${reason}`,
    });
  }

  private isAwaiting(state: RunState, requestId: RequestId): boolean {
    return (
      state.status === RunStatus.AwaitingApproval &&
      state.approval !== null &&
      state.approval.requestId === requestId
    );
  }

  private requireTool(name: string): Tool {
    const tool = this.tools.get(name);
    if (tool === undefined) {
      throw new UnknownToolError(name);
    }
    return tool;
  }

  private async recordAudit(
    runId: RunId,
    type: AuditEventType,
    data: Readonly<Record<string, unknown>>,
  ): Promise<void> {
    await this.audit.record({ runId, type, at: this.clock.now().toISOString(), data });
  }
}

/** Strip the handler so only the model-visible definition is sent to the provider. */
function toDefinition(tool: Tool): ToolDefinition {
  return {
    name: tool.name,
    description: tool.description,
    parameters: tool.parameters,
    risk: tool.risk,
  };
}

/** Render a tool result as a string for the tool message. */
function serialize(result: unknown): string {
  return typeof result === "string" ? result : (JSON.stringify(result) ?? "null");
}
